### PR TITLE
Reporting refactor

### PR DESCRIPTION
**Pull request by paulirish**
### Refactor of report loading

**Before**: Background page spawned a tab with `report-loader`, which used `chrome.runtime` API to tell background page that it was loaded. Then BG page passes over the `results` object. The new tab has all of handlebars.js locally and so it (`report-loader`) tells `report-generator` to template out the report, which it then  replaces the `innerHTML` of _its own document_ with the new stuff.

**After**: Background page takes results, uses the `report-generator` (still in the bg page) to template. Once complete, it spawns a new tab and set's the content of it (via blob url) to be the static report.

Sidenote: We can now trivially offer a "download report as HTML" option. :) And it'll be ~120kb instead of ~390kb from when `report-generator` sat inside the page.

### Changes:

* Report is generated into a static page, available by a blob URL. This will be used for both extension and in the devtools panel.
* Dropped the loading spinner of the report, as report generation takes ~100ms. Also now, the generation is done before the tab is opened (in the extension's case), so NBD.
* Nuked `report-loader` and cleaned up some places where we duplicated things in report in `/core` and `/extension`
* Error reporting (of handlebars compile failures) moved into `report-generator` and given a _nicer_ looking template
* Browserify of `lighthouse-background` split apart from the other files we browserify; 1) for clarity and 2) so we can run them in sequence.

### Blob url screenshot
image

### Error page screenshot
image

### CODE REVIEW

File: lighthouse-extension/app/src/lighthouse-background.js [hunk 1]
```diff
@@ -92,6 +83,28 @@ window.runLighthouseInWorker = function(port, url, options, requestedAudits) {
...
+
+  const reportGenerator = new ReportGenerator();
+  let html;
+  try {
+    html = reportGenerator.generateHTML(results);
+  } catch (err) {
+    html = reportGenerator.renderException(err, results);
+  }
+  const blob = new Blob([html], {type: 'text/html'});
+  const bloburl = window.URL.createObjectURL(blob);
```
Comment (ebidel) [inline lighthouse-extension/app/src/lighthouse-background.js]: `blobURL` to be consistent with above

File: lighthouse-extension/gulpfile.js [hunk 2]
```diff
@@ -94,61 +84,81 @@ gulp.task('chromeManifest', () => {
...
+    'app/src/lighthouse-background.js'
   ], {read: false})
     .pipe(tap(file => {
-      let bundle = browserify(file.path) // , {debug: true})
-      // Fix an issue with Babelified code that doesn't brfs well.
-      .transform('./fs-transform', {
+      let bundle = browserify(file.path); // , {debug: true})
+      bundle = applyBrowserifyTransforms(bundle);
+
+      // In the case of our lighthouse-core script, we've got extra work to do
```
Comment (ebidel) [inline lighthouse-extension/gulpfile.js]: work to do.

File: lighthouse-extension/gulpfile.js [hunk 3]
```diff
@@ -94,61 +84,81 @@ gulp.task('chromeManifest', () => {
...
+      // Do the additional transform to convert references to devtools-timeline-model
```
Comment (ebidel) [inline lighthouse-extension/gulpfile.js]: references to -> references of

File: lighthouse-extension/gulpfile.js [hunk 4]
```diff
@@ -14,7 +14,6 @@ const eslint = require('gulp-eslint');
 const livereload = require('gulp-livereload');
 const tap = require('gulp-tap');
 const zip = require('gulp-zip');
-const rename = require('gulp-rename');
 
 const audits = fs.readdirSync(path.join(__dirname, '../', 'lighthouse-core/audits/'))
```
Comment (ebidel) [inline lighthouse-extension/gulpfile.js]: note to me/you/all that we'll need to update this to hand the dbw directories. similarly with gatherers.

File: lighthouse-core/report/report-generator.js [hunk 5]
```diff
@@ -184,9 +183,23 @@ class ReportGenerator {
     return items;
   }
 
-  generateHTML(results, options) {
-    const inline = (options && options.inline) || false;
+  /**
+   * Creates the page describing any error generated while running generateHTML()
```
Comment (paulirish) [inline lighthouse-core/report/report-generator.js]: done

File: lighthouse-core/report/report-generator.js [hunk 6]
```diff
@@ -184,9 +183,23 @@ class ReportGenerator {
...
+      errStack: err.stack,
+      css: this.getReportCSS(),
+      results: JSON.stringify(results, null, 2)
+    });
+  }
 
+  /**
+   * Generates the Lighthouse report HTML
+   */
+  generateHTML(results) {
```
Comment (paulirish) [inline lighthouse-core/report/report-generator.js]: done

File: lighthouse-core/report/report-generator.js [hunk 7]
```diff
@@ -216,13 +229,14 @@ class ReportGenerator {
...
 
-    const template = Handlebars.compile(this.getReportHTML());
+    const template = Handlebars.compile(this.getReportTemplate());
     return template({
       url: results.url,
       lighthouseVersion: results.lighthouseVersion,
       generatedTime: this._formatTime(results.generatedTime),
-      css: this.getReportCSS(inline),
-      script: this.getReportJS(inline),
+      lhresults: JSON.stringify(results, null, 2),
```
Comment (paulirish) [inline lighthouse-core/report/report-generator.js]: mostly for readability. I'm dropping it in here for debuggability. 

> "This report looks a little whack.. Why the -1's? *Views source*.... Ohhh interesting stuff going on in the extended-info."

File: lighthouse-core/report/scripts/lighthouse-report.js [hunk 8]
```diff
@@ -67,6 +67,6 @@ LighthouseReport.prototype = {
   }
 };
 
-(function() {
+window.addEventListener('DOMContentLoaded', _ => {
   return new LighthouseReport();
```
Comment (paulirish) [inline lighthouse-core/report/scripts/lighthouse-report.js]: hah. wish I could. :)  

image

looks like this error doesnt exist in latest eslint/eslint-config-google, so i added a note to tweak once we upgrade.

File: lighthouse-core/report/templates/report-template.html [hunk 9]
```diff
@@ -16,18 +16,23 @@
...
-  </style>
+  <style>{{{ css }}}</style>
+  <script>{{{ script }}}</script>
 </head>
 <body>
-<div class="js-report report">
 
+<script>
+  // expose the original results
+  self.lhresults = {{ lhresults }};
```
Comment (paulirish) [inline lighthouse-core/report/templates/report-template.html]: pro. thx

File: lighthouse-extension/app/src/lighthouse-background.js (hunk 1 above)
Comment (paulirish) [inline lighthouse-extension/app/src/lighthouse-background.js]: done

File: lighthouse-extension/gulpfile.js (hunk 2 above)
Comment (paulirish) [inline lighthouse-extension/gulpfile.js]: rephrased.

File: lighthouse-extension/gulpfile.js (hunk 3 above)
Comment (paulirish) [inline lighthouse-extension/gulpfile.js]: sg

File: lighthouse-extension/gulpfile.js (hunk 4 above)
Comment (paulirish) [inline lighthouse-extension/gulpfile.js]: 👍

**Comment by paulirish** [comment]
> Cool. This looks like mostly prep for later? 

Following report creation was fairly circuitous. 

**Before**: Background page spawned a tab with report loader, which used chrome.runtime to tell background page that it was loaded. Then BG page passes over the results object. The new tab has all of handlebars.js locally and so it (report-loader) tells report-generator to template out the report, which it then  replaces the `innerHTML` of its own document with the new stuff.

**After**: Background page takes results, uses the report generator (still in the bg page) to template. Once complete, it spawns a new tab and set's the content of it (via blob url) to be the static report.

Just realized we can now offer a "download report as HTML" option trivially now. :) And it'll be ~120kb instead of ~390kb from when report-generator sat inside the page.

> Can you post a screenshot of the error page?

aye. two screenshots added up top.

File: lighthouse-core/report/templates/report-template.html (hunk 9 above)
Comment (paulirish) [inline lighthouse-core/report/templates/report-template.html]: Added a test assertion for this.

File: lighthouse-core/gather/driver.js [hunk 10]
```diff
@@ -423,7 +423,11 @@ class Driver {
     // Disable any domains that could interfere or add overhead to the trace
     return this.sendCommand('Debugger.disable')
       .then(_ => this.sendCommand('CSS.disable'))
-      .then(_ => this.sendCommand('DOM.disable'))
+      .then(_ => {
```
Comment (brendankenny) [inline lighthouse-core/gather/driver.js]: This is already on master. Need to rebase?

File: lighthouse-core/report/report-generator.js [hunk 11]
```diff
@@ -184,9 +183,28 @@ class ReportGenerator {
     return items;
   }
 
-  generateHTML(results, options) {
-    const inline = (options && options.inline) || false;
+  /**
+   * Creates the page describing any error generated while running generateHTML()
+   * @param {Exception} err Exception thrown from generateHTML
```
Comment (brendankenny) [inline lighthouse-core/report/report-generator.js]: `@param {!Error}`

File: lighthouse-extension/gulpfile.js (hunk 4 above)
Comment (brendankenny) [inline lighthouse-extension/gulpfile.js]: could we switch this to using `Runner.getAuditList()` (and `Runner.getGathererList` below)? That will keep these up to date as those change

File: lighthouse-extension/gulpfile.js [hunk 12]
```diff
@@ -94,61 +84,82 @@ gulp.task('chromeManifest', () => {
   .pipe(gulp.dest('dist'));
 });
 
-gulp.task('browserify', () => {
+function applyBrowserifyTransforms(bundle) {
+  // Fix an issue with Babelified code that doesn't brfs well.
```
Comment (brendankenny) [inline lighthouse-extension/gulpfile.js]: I don't remember what this comment means, but maybe remove since we don't use babel anymore?

**Comment by brendankenny** [comment]
all review issues addressed. merging for @paulirish

**Pull request merged**